sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala

```diff
@@ -34,6 +35,36 @@ import org.apache.spark.sql.internal.SQLConf.StoreAssignmentPolicy
 import org.apache.spark.sql.types.{ArrayType, DataType, DecimalType, IntegralType, MapType, StructType}
 
 object TableOutputResolver {
+
+  def resolveVariableOutputColumns(
+      expected: Seq[Expression],
+      query: LogicalPlan,
+      conf: SQLConf): LogicalPlan = {
+
+    if (expected.size != query.output.size) {
+      throw new AnalysisException(errorClass = "ASSIGNMENT_ARITY_MISMATCH",
+        messageParameters = Map("numTarget" -> expected.size.toString,
+          "numExpr" -> query.output.size.toString))
+    }
+
+    val resolved: Seq[NamedExpression] = {
+      query.output.zip(expected.asInstanceOf[Seq[VariableReference]]).map {
+        case (inputCol, expected) =>
+        if (DataTypeUtils.sameType(inputCol.dataType, expected.dataType)) {
+          inputCol
+        } else {
+          Alias(cast(inputCol, expected.dataType, conf, expected.varName), expected.varName)()
```

Comment:
shall we respect the store assignment policy, like INSERT does?

Reply:
Yes we shall, how do we do that?

Reply:
OK, I have changed it to ALWAYS use store assign and ANIS mode. Given that it's new that's defensible.